best-practices/multi-column-index-optimization.md

```diff
@@ -0,0 +1,263 @@
+
+Without these optimizations, query performance in large TiDB databases can degrade quickly. Full table scans and inadequate filtering can turn milliseconds into minutes. Additionally, excessive memory use can lead to out-of-memory (OOM) errors, especially in constrained environments. TiDB’s targeted approach ensures only relevant data is accessed. This keeps latency low and memory usage efficient—even for the most complex queries. Let’s explore how TiDB makes it happen.
+
+## Background: Multi-Column Indexes
+
+Indexes are a powerful tool for improving query performance by avoiding the need to scan all rows in a table. Let’s take an example of a rental listings table defined as follows. In this example, each listing contains a unique ID, city, number of bedrooms, rent price, and availability date:
+
+```sql
+CREATE TABLE listings (
+    listing_id INT PRIMARY KEY AUTO_INCREMENT,
+    city VARCHAR(100) NOT NULL,
+    bedrooms INT NOT NULL,
+    price DECIMAL(10, 2) NOT NULL,
+    availability_date DATE NOT NULL
+);
+```
+
+Suppose this table has 20 million listings across the United States. If we want to find all listings with a price under $2,000, we could add an index on the price column. This index would allow the optimizer to filter out rows, scanning only the range [ -inf, 2000.00 ). This helps reduce the search to about 14 million rows (assuming 70% of rentals are priced above $2,000). In the query execution plan, TiDB would perform an index range scan on price. This limits the need for a full table scan and improves efficiency.
+
+```sql
+"Q1" 
+explain FORMAT=BRIEF
+    SELECT * FROM listings WHERE price < 2000;
+-----+------------------------------------------------------+--------------------------
+| id  task                | access object                        | operator info   
++-----------------------------+---------+-----------+-----------------------------------
+| IndexLookUp             | root                                 | 
+| ├─IndexRangeScan(Build) | table:listings,index:price_idx(price)| range:[-inf,2000.00) 
+| └─TableRowIDScan(Probe) | table:listings                       | 
++-----------------------------+---------+-----------+-----------------------------------
+```
+
+While this filter improves performance, it may still return a large number of rows. This isn’t ideal for a user looking for more specific listings. Adding filters, such as specifying the city, number of bedrooms, and a maximum price, narrows the results significantly. For example, a query to find two-bedroom listings in San Francisco under $2,000 is more useful, likely returning only a few dozen rows.
```

Comment:
```suggestion
While this filter improves performance, it might still return a large number of rows. This is not ideal for a user looking for more specific listings. Adding filters, such as specifying the city, number of bedrooms, and a maximum price, narrows the results significantly. For example, a query to find two-bedroom listings in San Francisco under $2,000 is more useful, likely returning only a few dozen rows.
```